features/netsocket/Socket.h

```diff
@@ -225,6 +228,9 @@ class Socket {
     mbed::Callback<void()> _event;
     mbed::Callback<void()> _callback;
     rtos::Mutex _lock;
+    uint32_t _multicast_addr_count;
+    uint32_t _multicast_addr_registry;
+    SocketAddress  _socket_multicast_register[SOCKET_MAX_MULTICAST_MEMBERSHIPS];
```

Comment:
An array feels a little memory inefficient, especially if it was actually the multicast request with both group and interface address. Most sockets won't be doing any joins at all. 

Maybe have these fields in a malloced sublock you allocate on first join? Although I know some disfavour dynamic allocation.

Reply:
Agree, I wasn't sure what the stance on dynamic memory allocation was compared to static allocation in mbed generally so went with this option. Completely open to changing it.